sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -280,13 +280,20 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
     results.toArray
   }
 
+  private[spark] def executeCollectIterator(): (Long, Iterator[InternalRow]) = {
+    val countsAndBytes = getByteArrayRdd().collect()
```

Comment:
This still fetches all the compressed rows to the driver, before building the hashed relation. Ideally, you should fetch the rows from executors incrementally.

Reply:
I agree, but this is better than before. This only keeps all of the rows in memory compressed, and then streams through the compressed blocks. Before this patch, the rows are copied into a buffer per row while holding the compressed blocks, so it held the rows compressed and uncompressed at the same time. The uncompressed rows are what this fixes, we can follow up with something better to stream through blocks from executors.